### ISSUE TITLE
Inspire coverages and WCS

### ISSUE THREAD

**Issue by apfelnymous** [INSPIRE-MIF/helpdesk#22]
Hello I wanted to ask about the official status for the approach of having a wcs download service and providing Inspire feature data like OrthoimageCoverage in the metadata section of the inspire parent feature (like RectifiedGridCoverage).

Has there a request been made yet ? Anyone knows ?

Kind regards
János

**Comment by KathiSchleidt** [comment]
Hi János,

Coverages are a bit tricky in INSPIRE, so let me start from the top and work my way down :)
WCS has long been accepted as an INSPIRE Download Service, see the [Technical Guidance Doc for WCS in INSPIRE](https://inspire.ec.europa.eu/id/document/tg/download-wcs)

The tricky bit are the INSPIRE Coverage based models, thus the trick of shifting this additional information to a dataType provided in the Coverage Metadata. At present, our approach has [INSPIRE Candidate Good Practice ](https://inspire.ec.europa.eu/good-practice/ogc-compliant-inspire-coverage-data-and-service-implementation) status, no info at present when this will be confirmed (so goes from Candidate to Accepted)

The basic information on this is available from our [Webinar on INSPIRE and Coverages](https://inspire.ec.europa.eu/good-practice/ogc-compliant-inspire-coverage-data-and-service-implementation), the main point for information on this work a github repository at:
https://github.com/INSPIRE-MIF/gp-coverage-encoding

:)

Kathi

**Comment by apfelnymous** [comment]
Hi Kathi,

thank you for giving me the update. :)

We currently run this with mapserver.
https://github.com/wetransform-os/MapServer/wiki/INSPIRE-WCS-Extensions-for-EL-and-OI

I see rasdaman supports a request to return the gml/xml part only.
Do you think that this is essential overall ?

Mapserver supports a combination only so far.

**Comment by KathiSchleidt** [comment]
To my understanding, DescribeCoverage is an integral part of WCS

Counter-Question - does Mapserver allow you to add the additional INSPIRE XML to the metadata element of the coverage?

PS: please bear in mind that we're still awaiting final MIG approval for this modification to enable provision of INSPIRE requirements via WCS, thus, if you have an interest in utilizing WCS within INSPIRE, please get back to your national representatives (both MIG and MIG-T, as seems we get to go for yet another round with MIG-T on this, not quite clear what the process is). Otherwise there's a good chance that this work will be stopped

**Comment by HerzovanderWal** [comment]
Is it possible that these EU driven programms could take the enormous change by this technacly tiny improvement in sharing coverages? 

What I do know about this proposal is that a possible results is the data providing by Copernicus, EMODnet and/or SeaDatanet. The issue there is how do I keep my datastore uptodate! Sharing coverage with this tiny solution on Geotiff and BAG can help to exchange coverage data with there providers like Ministry of Infrastructure and Water Management in the Netherlands. There is another gamechanging aspect on this way of data sharing: It will give data-analysts and bigdata providers and users a boost in the technology and analyzing this coverage data. 

I would like to thank the specialist that they have recognised the GAP and shared the solution with us. I hope that the EU driven programms sees this improvement by this technacly tiny improvement in sharing coverages and help software developers to improve there tooling.
If you are interested in financing the implementation I think this picture of the forces can help:  
![image](https://user-images.githubusercontent.com/33277602/114377947-80e03680-9b87-11eb-80ae-1995c1cc8dfa.png)

**Comment by jescriu** [comment]
Thank you for your view, Herzo!

**Comment by KathiSchleidt** [comment]
@apfelnymous just saw your validation post on WCS and of course interrogated your services - noticed that you're currently missing the following 2 bits:
the INSPIRE extensions 

- required for all download services. GeoSolutions is currently extending to GeoServer, have a nice overview at:
  - https://geoserver.geo-solutions.it/educational/en/inspire/inspire_ext.html#extended-wcs-capabilities
- the additional bits of the INSPIRE data models, that we've been proposing the good practice on for providing this, see
  - https://inspire.ec.europa.eu/good-practice/ogc-compliant-inspire-coverage-data-and-service-implementation

**Comment by hogredan** [comment]
I would like to take the opportunity here to thank again the 'coverage experts' for the good work already done! From a German perspective it is very important to move the INSPIRE Good Practice on Coverages forward. We don't see any other way of providing coverage data INSPIRE and ISO-/OGC-compliant.

**Comment by KathiSchleidt** [comment]
Dear Potential Coverage Providers,

I'm aware of various activities in extending existing SW solutions to cover all INSPIRE requirements to WCS, in addition to the extension to the Capabilities page as specified in the [Technical Guidance for the implementation of INSPIRE Download Services using Web Coverage Services (WCS)](https://inspire.ec.europa.eu/id/document/tg/download-wcs) also providing support for the creation and inclusion of the additional INSPIRE data in the Coverage metadata element. 

I've repeatedly requested these SW providers to publicly announce this work, as it seems that the final sticking point in the approval of this Good Practice pertains to the lack of SW support as addressed yet again in last weeks MIG-T meeting. Not sure if this is due to non-disclosure agreements with their clients, or if they prefer to wait until this Good Practice is approved - either way, without public indication of these developments, all our hard work may yet come to nought :(

Thus - if you are working on implementing the necessary requirements in WCS, or have contracted this work, please be so kind as to indicate your activities here, so we can all get on with actually implementing this stuff!

:)

Kathi

**Comment by apfelnymous** [comment]
> 
> 
> To my understanding, DescribeCoverage is an integral part of WCS
> 
> Counter-Question - does Mapserver allow you to add the additional INSPIRE XML to the metadata element of the coverage?
> 
Not until the modification project I linked in my first comment.
The resulting file from a getcoverage multipart request looks like this example(can be opened wih an editor, inspire extension starts at line 79):

[getcoverage-multipart-xml](https://geoportal.saarland.de/gdi-sl/inspirewcs?map=/home/lvgl/inspire_oi.map&SERVICE=WCS&VERSION=2.0.0&REQUEST=GetCoverage&COVERAGEID=sl_inspire_oi_dop20&SUBSET=x(2524000,2525000)&SUBSET=y(5442000,5443000)&RESOLUTION=x(0.4)&RESOLUTION=y(0.4)&CRS=EPSG:31466&RESPONSE_CRS=EPSG:31466&WIDTH=10&HEIGHT=10&FORMAT=GEOTIFF_RGB&MEDIATYPE=multipart/related)

We currently need this functionality for Orthoimagery and Elevation.

**Comment by KathiSchleidt** [comment]
@apfelnymous any insights when this MapServer functionality will be available, if it will also be in the general branch, or only a customer specific one?

**Comment by apfelnymous** [comment]
> 
> 
> @apfelnymous any insights when this MapServer functionality will be available, if it will also be in the general branch, or only a customer specific one?

We requested this to be merged into the general branch.
Last time I asked about when the changes will be pulled it was unknown, I will keep updating you.

**Comment by simboss** [comment]
@KathiSchleidt just chiming in from the GeoServer perspective.

Work has been done in the past from the [British Geological Survey](https://www.bgs.ac.uk/) to extend the [INSPIRE Extension](https://docs.geoserver.org/latest/en/user/extensions/inspire/index.html) to have GeoServer comply with the [WCS requirements for INSPIRE](https://docs.geoserver.org/latest/en/user/extensions/inspire/using.html#extended-wfs-and-wcs-configuration).

We are currently doing a review to assess the missing bits; we shall be able to put together a report in the upcoming weeks.

**Comment by KathiSchleidt** [comment]
@simboss will this also be supporting the requirements from the Coverage Good Practice, the inclusion of the necessary attributes for the INSPIRE Coverages via the Coverage metadata element? Or only the basic extensions foreseen under the WCS requirements?

**Comment by simboss** [comment]
What we are doing at the moment is an assessment, we'll try to cover as much as possible.

**Comment by heidivanparys** [comment]
The good practice on coverages was endorsed at the [13th MIG meeting](https://webgate.ec.europa.eu/fpfis/wikis/pages/viewpage.action?pageId=778175536), can this issue be closed?

**Comment by KathiSchleidt** [comment]
Both issues can now be closed, as long as JRC is aware that the XSDs they provide do not suffice. The GP currently depends on the XSD files available from datacove.eu, at present available for EL, OI and LC, still missing for the other coverage themes

**Comment by heidivanparys** [comment]
I am not into the details of the Coverage Good Practice, but if certain XML schemas need to be added to the INSPIRE XML Schema server, see the dedicated GitHub repository on https://github.com/INSPIRE-MIF/application-schemas/.

The limitations are described in the Good Practice as far as I can see.

**Comment by KathiSchleidt** [comment]
@heidivanparys I trust that you will assure that the necessary issues for the application schemas will be created, or do you expect MS to create issues on a theme by theme basis as they implement? (seems like a fairly inefficient manner of creating the schema files piecemeal, and not really supportive to MS implementing)
Also, while I understand the satisfaction behind closing issues, I'd have expected the necessary followup to be done BEFORE this issue is closed

**Comment by heidivanparys** [comment]
At the end of [the webinar presenting the good practice for coverages](https://www.youtube.com/watch?v=WrjzYCeLpfw), it was agreed to set up a dedicated repository to host discussions related to that topic. For those following this GitHub issue, please make sure to watch the https://github.com/INSPIRE-MIF/gp-coverage-encoding repository, so you get notified of any new discussions there, and you are also very welcome to start new discussions there.

In section 5.2 in the [discussion paper presenting the good practice itself]( https://webgate.ec.europa.eu/fpfis/wikis/download/attachments/527439698/INSPIRE_COVERAGES_Discussion_document_v1r0.pdf?api=v2) it says:

> As explained in the Section 4, this discussion document proposes to:
> 
> 1. Directly adopt OGC CIS 1.0 coverage implementation schema as default encoding for INSPIRE coverage data models:
> http://schemas.opengis.net/cis/1.0/coverage.xsd
> 2. Define INSPIRE coverage metadata schemas in order to provide all the additional elements required by the Directive (INSPIRE extensions).
> The draft proposed Elevation and Orthoimagery schemas for providing INSPIRE coverage metadata based on the present good practice are accessible through the next links:
>   - Elevation: https://schema.datacove.eu/ElevationGridCoverageMetadata.xsd
>   - Orthoimagery: https://schema.datacove.eu/OrthoimageryMetadata.xsd
> NOTE: Despite not being related to the example themes showed in this discussion document, there is also available a proposed schema for the Land cover theme: http://test.datacove.eu/LandCoverRasterMDExt.xsd
> The previous sample schemas should be revised with a minimum effort in order check consistency and completeness of the all the properties included in the INSPIRE Implementing Rule for the Interoperability of Data sets and Services for the above-mentioned themes.
> Ideally JRC should be hosting the final schemas for their custody and usage in the long term, in case of adopting the present good practice.

The [INSPIRE Good Practices](https://inspire.ec.europa.eu/portfolio/good-practice-library) are community driven, so the community behind the good practice on coverages should drive that process.

The governance of the INSPIRE XML schemas is done via https://github.com/INSPIRE-MIF/application-schemas/, see that repo and the [presentation](https://webgate.ec.europa.eu/fpfis/wikis/download/attachments/712606761/MIG-T65_GitHub_INSPIRE_artefacts.pdf?version=1&modificationDate=1618560336698&api=v2) on that topic from the 65th MIG-T meeting for more information.

**Comment by KathiSchleidt** [comment]
Many thanks for the kind clarification, fear I'd misinterpreted the following statement from page 2 of the "Technical Guidance for INSPIRE Download Services using WCS":

`This technical guidance shows how the operations required by the [INS NS] for download services can be mapped to the WCS 2.0 standard. A **second document will be provided** to show how the data specifications that have a requirement to provide coverage data, might encode their data to provision it through an INSPIRE conformant download service based on WCS as documented in this guidance.`

I had assumed that both documents would be provided by the EC, my mistake, apologies.
